### PR TITLE
Campbell R. Harvey and Yan Liu "Backtesting" module

### PR DESCRIPTION
**Pull request by PanPip**
# Description

This is the implementation of the Haircut Sharpe Ratios and Profit Hurdle algorithms described in the work of Campbell R. Harvey and Yan Liu "Backtesting". The code is built in the form of a class.

The unit test and Sphinx documentation are also provided with the code. The notebook with algorithms explanation and examples of use are in the H&T research repository.

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

A unit test for class methods outputs and coverage is provided.

- [x] test_backtests.py

**Test Configuration**:
* Windows 10
* PyCharm

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/source/implementations/backtest_overfitting.rst [hunk 1]
```diff
@@ -0,0 +1,146 @@
+.. _implementations-backtest_overfitting:
+
+============================================================
+Backtest Overfitting by Campbell R. Harvey and Yan Liu
```
Comment (Jackal08) [inline docs/source/implementations/backtest_overfitting.rst]: The original paper is just called ```Backtesting``` Lets remove Overfitting from the title. Unless I missed something?

File: mlfinlab/__init__.py [hunk 2]
```diff
@@ -19,9 +19,11 @@
 import mlfinlab.ensemble as ensemble
 import mlfinlab.portfolio_optimization as portfolio_optimization
 import mlfinlab.clustering as clustering
+import mlfinlab.backtest_statistics.backtests as backtests
```
Comment (Jackal08) [inline mlfinlab/__init__.py]: This was a good idea given the paper was called Backtest.

File: mlfinlab/backtest_statistics/backtests.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+        sr_adj[3], haircut[3] = self._sharpe_ratio_haircut(p_val_adj[3], monthly_obs, sr_annual)
+
+        results = np.array([p_val_adj,
+                            sr_adj,
+                            haircut])
+
+        return results
+
+    def profit_hurdle(self, num_mult_test, num_obs, alpha_sig, vol_anu, rho):
+        # pylint: disable=too-many-locals
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/backtests.py]: Be careful when you start adding the ignore linter strings. Usually the errors they raise help us to write better code. Only ignore them when you make an overriding decision.

File: mlfinlab/backtest_statistics/backtests.py [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+        '''
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/backtests.py]: All these docstrings should please be with double quotes and not single quotes. 

"""

\"""

File: mlfinlab/backtest_statistics/backtests.py [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+                # And the corresponding t-statistic from p-value
+                tstat_b = ss.norm.ppf((1 - (p_val[0] + p_chosen) / 4), 0, 1)
+
+        return tstat_b
+
+    def haircut_sharpe_ratios(self, sampling_frequency, num_obs, sharpe_ratio, annualized,
+                              autocorr_adjusted, rho_a, num_mult_test, rho):
+        # pylint: disable=too-many-locals
+        """
+        Calculate Sharpe ratio adjustments due to testing multiplicity
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/backtests.py]: Grammar, please. Missing full stops.

File: docs/source/implementations/backtest_overfitting.rst (hunk 1 above)
Comment (PanPip) [reply docs/source/implementations/backtest_overfitting.rst]: It is. I was still thinking about the naming. Will change.

File: mlfinlab/backtest_statistics/backtests.py (hunk 3 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/backtests.py]: In this case, I had more variables for the purpose of better code description. I believe the maximum allowed by pylint is 20, whereas I had 22. I believe I've already decreased them to an acceptable level here.

File: mlfinlab/backtest_statistics/backtests.py (hunk 4 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/backtests.py]: Will change

File: mlfinlab/backtest_statistics/backtests.py (hunk 5 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/backtests.py]: Will change

File: mlfinlab/backtest_statistics/backtests.py [hunk 6]
```diff
@@ -0,0 +1,546 @@
+# pylint: disable=missing-module-docstring
+
```
Comment (aditya1702) [inline mlfinlab/backtest_statistics/backtests.py]: Remove space here.

File: mlfinlab/backtest_statistics/backtests.py [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+        # The first row of the correlation matrix: [1, rho, rho, .., rho]
+        correlation_vector = np.insert(rho * np.ones((1, n_trails - 1)), 0, 1)
+        # Correlation matrix created from the vector by expanding it
+        correlation_matrix = linalg.toeplitz(correlation_vector)
+        # Vector with mean of simulated returns - zeros
+        mean = np.zeros(n_trails)
+
+        # Creating a sample from a multivariate normal distribution as returns simulations
+        # Covariance matrix - Created from correlation matrix multiplied by monthly volatility and adjusted
+        covariance_matrix = correlation_matrix * (monthly_volatility ** 2 / n_obs)
```
Comment (aditya1702) [inline mlfinlab/backtest_statistics/backtests.py]: Please remove spaces between comments. That way it looks easy to read the code. Do this for all parts of the code. Something like this:

```
# Creating a sample from a multivariate normal distribution as returns simulations
# Covariance matrix - Created from correlation matrix multiplied by monthly volatility and adjusted
covariance_matrix = correlation_matrix * (monthly_volatility ** 2 / n_obs)

 # Result - n_simulations rows with n_trails inside
 shock_mat = np.random.multivariate_normal(mean, covariance_matrix, n_simulations)
```

File: mlfinlab/backtest_statistics/backtests.py [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+        """
+        Estimates the parameters used to generate the distributions in _sample_random_multest - the HLZ model.
+
+        Based on the work of HLZ, the pairwise correlation of returns is used to estimate the probability (prob_zero_mean),
+        total number of trials (n_simulations) and (lambd) - parameter of the exponential distribution. Levels and
+        parameters taken from the HLZ research.
+
+        :param rho: (float) Average correlation coefficient between strategy returns
+        :return: (np.array) Array of parameters
+        """
```
Comment (aditya1702) [inline mlfinlab/backtest_statistics/backtests.py]: Add space after the docstrings.

File: mlfinlab/backtest_statistics/backtests.py [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+    @staticmethod
+    def _monthly_observations(num_obs, sampling_frequency):
+        """
+        Calculates the number of monthly observations based on sampling frequency and number of observations.
+
+        :param num_obs: (int) Number of observations used for modelling
+        :param sampling_frequency: (str) Sampling frequency of returns
+                                   ['D','W','M','Q','A'] = [Daily, Weekly, Monthly, Quarterly, Annual]
+        :return: (np.float64) Number of monthly observations
+        """
```
Comment (aditya1702) [inline mlfinlab/backtest_statistics/backtests.py]: Space after docstring. Do this for all other functions as well

File: mlfinlab/backtest_statistics/backtests.py [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+        # Result is [rho, n_simulations, prob_zero_mean, lambd]
+        parameters = self._parameter_calculation(rho)
+
+        # Needed number of trails inside a simulation with the check of (num_simulations >= num_mul_tests)
+        num_trails = int((np.floor(num_mult_test / parameters[1]) + 1) * np.floor(parameters[1] + 1))
+        # Generating a panel of t-ratios (of size self.simulations * num_simulations)
+        t_sample = self._sample_random_multest(parameters[0], num_trails, parameters[2], parameters[3],
+                                               self.simulations)
+
+        # Holm method
```
Comment (aditya1702) [inline mlfinlab/backtest_statistics/backtests.py]: I see that the code below for Holm and BHY method is almost the same except for the calls to self._holm_method_returns() and self._bhy_method_returns(). You can try to minimise the redundant code here by adding them in a single FOR loop.

File: mlfinlab/backtest_statistics/backtests.py (hunk 6 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/backtests.py]: Will do

File: mlfinlab/backtest_statistics/backtests.py (hunk 7 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/backtests.py]: Will do

File: mlfinlab/backtest_statistics/backtests.py (hunk 8 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/backtests.py]: Will do

File: mlfinlab/backtest_statistics/backtests.py (hunk 9 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/backtests.py]: Will do

File: mlfinlab/backtest_statistics/backtests.py (hunk 10 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/backtests.py]: Did both methods in one FOR loop. Thank you.

File: mlfinlab/tests/test_backtest_statistics.py [hunk 11]
```diff
@@ -35,6 +34,7 @@ def setUp(self):
         """
         Set the data for tests.
         """
+
         project_path = os.path.dirname(__file__)
```
Comment (Jackal08) [inline mlfinlab/tests/test_backtest_statistics.py]: Do we need this? I think you can reference the local dir as such: ```./test_data/dollar_bar_sample.csv```

**Pull request merged**